File: sdk/python/examples/subscribe.py
"""Subscribe to new blocks over the JSON-RPC WebSocket endpoint.

Hara Registry exposes ``wss://rpc.ledger.haratrust.io/ws`` for ``eth_subscribe``
(newHeads / logs). This uses web3.py's WebSocket provider.

Requires web3 with websockets support (``pip install 'web3[ws]'`` or just
``websockets``). Run:

    python examples/subscribe.py
"""

from __future__ import annotations

import asyncio

from web3 import AsyncWeb3
from web3.providers.persistent import WebSocketProvider

from hara_registry import config


async def watch_new_heads() -> None:
    async with AsyncWeb3(WebSocketProvider(config.RPC_WS_URL)) as w3:
        print(f"connected to {config.RPC_WS_URL}; chain id {await w3.eth.chain_id}")
        sub_id = await w3.eth.subscribe("newHeads")
        print(f"subscribed to newHeads ({sub_id}); waiting for blocks ...")

        async for message in w3.socket.process_subscriptions():
            head = message["result"]
            number = int(head["number"], 16)
            print(f"new block #{number}  hash={head['hash'].hex()}")


def main() -> None:
    try:
        asyncio.run(watch_new_heads())
    except KeyboardInterrupt:
        print("\nstopped.")


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/subscribe.ts
/**
 * Subscribe to new block headers over WebSocket.
 *
 *   npx tsx examples/subscribe.ts
 *
 * Uses the public WS endpoint wss://rpc.ledger.haratrust.io/ws (no auth).
 */
import { createPublicClient, webSocket } from "viem";
import { ENDPOINTS, haraChain } from "@hara/registry-sdk";

async function main() {
  const client = createPublicClient({
    chain: haraChain,
    transport: webSocket(ENDPOINTS.ws),
  });

  console.log("Subscribing to newHeads on", ENDPOINTS.ws);
  const unwatch = client.watchBlocks({
    emitMissed: false,
    onBlock: (block) => {
      console.log(`block #${block.number}  ${block.hash}  txs=${block.transactions.length}`);
    },
    onError: (err) => console.error("subscription error:", err),
  });

  // Run for ~30s then stop (instant-finality ~2s blocks → ~15 blocks).
  setTimeout(() => {
    unwatch();
    console.log("Unsubscribed.");
    process.exit(0);
  }, 30_000);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
